tests/test_gates_gates.py

```diff
@@ -1574,23 +1562,6 @@ def test_controlled_unitary(backend):
     backend.assert_allclose(final_state, target_state)
 
 
-def test_controlled_unitary_matrix(backend):
-    nqubits = 2
-    initial_state = random_statevector(2**nqubits, backend=backend)
-
-    matrix = np.random.random((2, 2))
-    gate = gates.Unitary(matrix, 1).controlled_by(0)
-
-    target_state = apply_gates(backend, [gate], nqubits, initial_state)
-
-    u = backend.control_matrix(gate)
-    u = backend.cast(u, dtype=u.dtype)
-
-    final_state = np.dot(u, initial_state)
-
-    backend.assert_allclose(final_state, target_state)
```

Comment:
Instead of completely dropping this test, I would suggest to port it to the new interface you propose above for calculating the matrix of the controlled gate using `circuit.unitary()`. You could also do this for various examples, such as a `Unitary` gate and the example you provided in the docstring above:
```py
nqubits = 5
circuit = Circuit(nqubits)
circuit.add(gates.SWAP(3, 4).controlled_by(0, 1, 2))
print(circuit.unitary()) # check if this is the matrix you expect
```
Switching the order of controls and targets will also reveal if the issue in my previous comment is true or not.

Reply:
Indeed, I tried a few examples and it seems to work as expected, thanks. It would still be good to have some test of the `circuit.unitary()` when the circuit contains a non-trivial `.controlled_by` gate, since we advertise that feature in the docs.